File: 733-flood-fill/flood-fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} color
 * @return {number[][]}
 */
var floodFill = function(image, sr, sc, color) {
    if (image[sr][sc] === color) return image

    fill(image, sr, sc, image[sr][sc], color)
    return image
};

const fill = (image, sr, sc, color, newColor) => {
    if (sr < 0 || sr >= image.length || sc < 0 || sc >= image[0].length || image[sr][sc] !== color) return

    image[sr][sc] = newColor
    fill(image, sr + 1, sc, color, newColor)
    fill(image, sr - 1, sc, color, newColor)
    fill(image, sr, sc + 1, color, newColor)
    fill(image, sr, sc - 1, color, newColor)
}

File: 733-flood-fill/flood-fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:           
        if image[sr][sc] == color:
            return image
        
        original_color = image[sr][sc]
        directions = [[-1,0],[1,0],[0,-1],[0,1]]
        rows, cols = len(image), len(image[0])

        queue = deque([(sr, sc)])
        image[sr][sc] = color

        while queue:
            row, col = queue.popleft()
            image[row][col] = color
            for dr, dc in directions:
                new_r, new_c = row + dr, col + dc
                if self.is_within_bounds(new_r, new_c, rows, cols) and image[new_r][new_c] == original_color:
                    # image[new_r][new_c] = color
                    queue.append((new_r, new_c))

        return image

    def is_within_bounds(self, r, c, rows, cols):
        return 0 <= r < rows and 0 <= c < cols
        